File: Silver/2016 Dec/moocast.py
def flood_fill(start, adj):
  print("NEW START", start)
  visited = set()
  reached = 0
  queue = [start]
  while len(queue) > 0:
    curr = queue.pop(0)
    visited.add(curr)
    reached += 1
    print("reached", curr)
    for i in adj[curr]:
      if i not in visited:
        queue.append(i)
        visited.add(i)
  return reached

from collections import defaultdict
import math

def Run(input, output):
  with open(input, 'r') as fin, open(output, "w") as fout:
    N = int(fin.readline())
    cows = []
    adj = defaultdict(lambda: set())

    for i in range(N):
      x, y, p = (int(j) for j in fin.readline().split())
      cows.append((x, y, p))

    for point1 in range(N):
      x1, y1, p1 = cows[point1]
      for point2 in range(point1 + 1, N):
        x2, y2, p2 = cows[point2]

        dist = math.sqrt( (x1 - x2)**2 + (y1 - y2)**2 )

        if p1 >= dist:
          adj[point1].add(point2)
        if p2 >= dist:
          adj[point2].add(point1)
    
    print(adj)
    
    ans = 1

    for point in range(N):
      reached = flood_fill(point, adj)
      if reached > ans:
        ans = reached
    
    fout.write("{}\n".format(ans))


Run("moocast.in", "moocast.out")

File: Silver/2016 Dec/moocast.cpp
#include <bits/stdc++.h>
#include <cmath>

#define ll long long
#define MAXN 210

using namespace std;

struct Cow {
	ll x, y, p;
};

ll N;
Cow cows[MAXN];
vector<ll> adj[MAXN];

ll floodfill(ll start, bool visited[MAXN]) {
	ll ans = 0;
	if (visited[start]) {
		return ans;
	}

	visited[start] = true;
	ans++;

	for (auto i: adj[start]) {
		if (!visited[i]) {
			ans += floodfill(i, visited);
		}
	}
	return ans;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	
	string fname = "moocast";
	freopen((fname + ".in").c_str(), "r", stdin);
	freopen((fname + ".out").c_str(), "w", stdout);
	
	cin >> N;

	for (ll i = 0; i < N; i++) {
		ll x, y, p;
		cin >> x >> y >> p;

		Cow cow; 
		cow.x = x;
		cow.y = y;
		cow.p = p;

		cows[i] = cow;
	}

	for (ll i = 0; i < N - 1; i++) {
		for (ll j = i + 1; j < N; j++) {
			double dist = sqrt(pow(cows[i].x - cows[j].x, 2) + 
							pow(cows[i].y - cows[j].y, 2));
			//cout << "dist between " << i << " " << j << " is " << dist << "\n";
			if (dist <= cows[i].p) {
				adj[i].push_back(j);
			}	
			if (dist <= cows[j].p) {
				adj[j].push_back(i);
			}
		}
	}

	/*for (ll i = 0; i < N; i++) {
		for (auto j: adj[i]) {
			cout << i << " can reach " << j << " ";
		}
		cout << "\n";
	}*/

	ll ans = 1;

	for (ll i = 0; i < N; i++) {
		bool visited[MAXN] = {false};
		ll ff = floodfill(i, visited);
		if (ff > ans) {
			ans = ff;
		}
	}

	cout << ans << "\n";

	return 0;
}